keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,853 @@
+```
+
+Somewhat related to this, there is currently no way for a
+NetworkPolicy implementation (ie, the network plugin) to indicate the
+status of NetworkPolicy processing to the user who created the policy.
+
+This KEP attempts to solve both of these problems.
+
+(Although it is not technically required that NetworkPolicy be
+implemented by the network plugin, this is almost universally the case,
+and so I will refer to NetworkPolicy implementations as "network
+plugins" through the rest of this document.)
+
+## Motivation
+
+### Goals
+
+- Provide a better mechanism for dealing with version skew with
+  respect to NetworkPolicy features. Specifically:
+
+    - Allow users to indicate explicitly what version of the
+      NetworkPolicy featureset they are making use of (eg, whether
+      they require support for combined `podSelector` and
+      `namespaceSelector`), so that network plugins can reliably
+      recognize when a NetworkPolicy makes use of a feature that the
+      NetworkPolicy implementation does not support.
+
+    - Allow NetworkPolicy implementations to indicate back to the user
+      when they are aware that they cannot correctly implement a given
+      NetworkPolicy.
+
+- Define rules for dealing with feature gates and alpha APIs in
+  NetworkPolicy that work well with the 3-way versioning split.
+
+- Allow network plugins to indicate when a NetworkPolicy has been fully
+  "programmed" into the cluster network, to allow clients (including
+  humans, scripts, and cluster components) to get more reliable behavior
+  when adding new policies.
+
+- (Maybe) Allow network plugins to provide some sort of global status
+  object with certain information about the implementation.
+
+### Non-Goals
+
+- Deprecating (or requiring) the use of feature gates to indicate the
+  stability / "certainty" of NetworkPolicy API additions; whether a
+  feature is API-stable or not is distinct from whether it is widely
+  implemented by plugins or not (though there is generally bidirectional
+  influence between the two).
+
+- Defining "metric-like" NetworkPolicy status information (eg, how long
+  a particular rule took to implement).
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Testing Features of Different NetworkPolicy Implementations
+
+As a Kubernetes developer (who might be named Jay), I want to create a
+NetworkPolicy test suite, and test the NetworkPolicy implementations of
+various network plugins, but I don't want to have to manually keep track
+of which plugins (and which versions of which plugins) are expected to
+correctly implement which NetworkPolicy features.
+
+#### Story 2 - More Reliable NetworkPolicy Test Cases
+
+As a developer, I want NetworkPolicy e2e test cases that work reliably,
+meaning I want the test code to be able to tell at what point a
+newly-created NetworkPolicy is expected to be in effect, so that I don't
+time out the test case before the network plugin has managed to process
+the policy.
```

Comment:
Is there a way to determine if a particular network plugin supports the status field? Just thinking about the e2e test cases in network_policy.go. Once this KEP is approved, I would imagine the e2e test would wait for the net policy status to be enforcing before spinning up the client pods to validate connectivity.  Ah... I see that section below... let me look there.